File: 780-max-chunks-to-make-sorted/max-chunks-to-make-sorted.cpp
class Solution {
public:
    unordered_map<int,int> d; 
    int dfs(int i , vector<int>& arr){
        if(i==arr.size()) return 0;
        if(d.find(i)!=d.end()) return d[i];
        d[i] =0;
        int req = 0,s=0;
        for(int j = i;j<arr.size();j++){
            req+=j;
            s+=arr[j];
            if(req == s) d[i] = max(d[i],1 + dfs(j+1,arr));
        }
        return d[i];
    }
    int maxChunksToSorted(vector<int>& arr) {
        return dfs(0,arr);
    }
};

File: 780-max-chunks-to-make-sorted/max-chunks-to-make-sorted.py
class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        
        d= {}
        def dfs(i):
            if i == len(arr):return 0
            if i in d:return d[i]
            d[i] = -len(arr)
            req = 0
            s = 0
            for j in range(i,len(arr)):
                s += arr[j]
                req += j
                if req == s:
                    d[i]  = max(d[i],1 + dfs(j+1))
            return d[i]
        
        return dfs(0)
